docs/guides/end-to-end-testing/testing-strategies/azure-active-directory-authentication.mdx

```diff
@@ -0,0 +1,261 @@
+      },
+    },
+    ({ password }) => {
+      cy.get('input[type="password"]').type(password, {
+        log: false,
+      })
+      cy.get('input[type="submit"]').click()
+      cy.get('#idBtn_Back').click()
+    }
+  )
+
+  // Ensure Microsoft has redirected us back to the sample app with our logged in user.
+  cy.url().should('equal', 'http://localhost:3000/')
+  cy.get('#welcome-div').should(
+    'contain',
+    `Welcome ${Cypress.env('aad_username')}!`
+  )
+}
+
+Cypress.Commands.add('loginToAAD', (username: string, password: string) => {
+  const log = Cypress.log({
+    displayName: 'Azure Active Directory Login',
+    message: [`🔐 Authenticating | ${username}`],
+    autoEnd: false,
+  })
+  log.snapshot('before')
+
+  loginViaAAD(username, password)
+
+  log.snapshot('after')
+  log.end()
+})
+```
+
+Now, we can use our `loginToAAD` command in the test. Below is our test to login
+as a user via Azure Active Directory and run a basic sanity check.
+
+```js
+describe('Azure Active Directory Authentication', () => {
+  beforeEach(() => {
+    // log into Azure Active Directory through our sample SPA using our custom command
+    cy.loginToAAD(Cypress.env('aad_username'), Cypress.env('aad_password'))
+    cy.visit('http://localhost:3000')
+  })
+
+  it('verifies the user logged in has the correct name', () => {
+    cy.get('#table-body-div td:contains("name") + td').should(
+      'contain',
+      `${Cypress.env('aad_name')}`
+    )
+  })
+
+  it('verifies the user logged in has the correct preferred name', () => {
+    cy.get('#table-body-div td:contains("preferred_username") + td').should(
+      'contain',
+      `${Cypress.env('aad_username')}`
+    )
+  })
+})
+```
+
+<DocsVideo src="https://vimeo.com/789095126"></DocsVideo>
+
+We now have working authentication and tests! But we are logging in before every
+test, which is not only time consuming, but also can lead to API rate limiting
+due to the number of requests.
+
+For this, we can refactor our login command to take advantage of
+[`cy.session()`](/api/commands/session) to store our logged in user's tokens
+and/or cookies so we don't have to reauthenticate before every test.
+
+```js
+Cypress.Commands.add('loginToAAD', (username: string, password: string) => {
+  cy.session(
+    `aad-${username}`,
+    () => {
+      const log = Cypress.log({
+        displayName: 'Azure Active Directory Login',
+        message: [`🔐 Authenticating | ${username}`],
+        // @ts-ignore
+        autoEnd: false,
+      })
+
+      log.snapshot('before')
+
+      loginViaAAD(username, password)
+
+      log.snapshot('after')
+      log.end()
+    },
+    {
+      validate: () => {
+        // this is a very basic form of session validation for this demo.
+        // depending on your needs, something more verbose might be needed
+        cy.visit('http://localhost:3000')
+        cy.get('#welcome-div').should(
+          'contain',
+          `Welcome ${Cypress.env('aad_username')}!`
```

Comment:
this feels like a funky use for the validate function. Whats it suppose to be doing here? If the idea is to get it to wait for the values to be ready before the session is serialized I would move the `cy.get('#welcome-div')` command to the bottom of the cy.session setup method.

Reply:
mainly the goal is to make sure the session is rehydrated correctly with the correct user logged in when the session is reinvoked. Moving to the end of the setup method method would mean it would only work the first time